kernel/kernel-impl/src/main/java/org/sakaiproject/messaging/impl/UserMessagingServiceImpl.java

```diff
@@ -127,7 +153,58 @@ public void init() {
 
         messaging = ignite.message(ignite.cluster().forLocal());
 
+        Security.addProvider(new BouncyCastleProvider());
+
         executor = Executors.newFixedThreadPool(20);
+
+        this.pushEnabled = serverConfigurationService.getBoolean("portal.notifications.push.enabled", false);
+
+        if (pushEnabled) {
+            String home = serverConfigurationService.getSakaiHomePath();
+            String publicKeyFileName = serverConfigurationService.getString(PUSH_PUBKEY_PROPERTY, "sakai_push.key.pub");
+            Path publicKeyPath = Paths.get(home, publicKeyFileName);
+            String privateKeyFileName = serverConfigurationService.getString("portal.notifications.push.privatekey", "sakai_push.key");
+            Path privateKeyPath = Paths.get(home, privateKeyFileName);
+
+            if (Files.exists(privateKeyPath) && Files.exists(publicKeyPath)) {
+                try {
+                    publicKey = String.join("", Files.readAllLines(Paths.get(home, publicKeyFileName)));
+                    String privateKey = String.join("", Files.readAllLines(Paths.get(home, privateKeyFileName)));
+                    pushService = new PushService(publicKey, privateKey);
+                    String pushSubject = serverConfigurationService.getString("portal.notifications.push.subject", "");
+                    pushService.setSubject(pushSubject);
+                } catch (Exception e) {
+                    log.error("Failed to setup push service: {}", e.toString());
+                }
+            } else {
+                ECNamedCurveParameterSpec parameterSpec = ECNamedCurveTable.getParameterSpec("prime256v1");
+
+                try {
+                    KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance("ECDH", BouncyCastleProvider.PROVIDER_NAME);
+                    keyPairGenerator.initialize(parameterSpec);
+
+                    KeyPair keyPair = keyPairGenerator.generateKeyPair();
+
+                    byte[] publicKey = Utils.encode((ECPublicKey) keyPair.getPublic());
+                    String publicKeyBase64 = Base64.getUrlEncoder().withoutPadding().encodeToString(publicKey);
+                    try (FileWriter fw = new FileWriter(publicKeyPath.toFile())) {
+                        fw.write(publicKeyBase64);
```

Comment:
what happens here with 5 nodes in a Sakai cluster?

Reply:
they should all get exactly the same key in their TOMCAT/sakai directory. The component will initialised on every node, so the same code should run on every node.

Reply:
@ottenhoff Do you think that is a bad strategy, storing a keypair on every node? Do you think a key server is a better idea, a single key server of some kind? That would add another security angle, of course, as we'd need to pull those keys across the network when formulating the push packets.

Reply:
seems like a mistake, yes. there should be one key set in sakai.properties

Reply:
Do you mean just specify the key basename, without the extension?

Reply:
@ottenhoff I want to get this merged - it's either this or short polling, seeing as the service worker approach is not working with anything other than Chrome. Each node has to have the keypair available to encrypt the push request and the same keypair will be generated with the same parameters, so each node will have the same keypair. I'm not sure what your concern is but I want to make sure I address it before merging.

Reply:
the private key placed on each node is fine with me